### PR TITLE
hevm: add SHA256, RIPEMD, IDENTITY, MODEXP, ECADD, ECMUL, ECPAIRING precompiles

### PR DESCRIPTION
**Pull request by livnev**
and also fixes some precompile-related bugs, refactors precompile handling.

**GeneralStateTests status**: 6015 passed, 3782 bad-balance, 815 failed.

Thanks @cmditch 😊.  This is now ready to merge IMO.

### CODE REVIEW

File: src/hevm/hevm.cabal [hunk 1]
```diff
@@ -12,7 +13,7 @@ description:
 homepage:
   https://github.com/dapphub/dapptools
 license:
-  AGPL-3
+  AGPL-3.0-only
```
Comment (rainbreak) [inline src/hevm/hevm.cabal]: Is this automated?

File: src/hevm/hevm.cabal [hunk 2]
```diff
@@ -1,3 +1,4 @@
+cabal-version: 2.2
```
Comment (rainbreak) [inline src/hevm/hevm.cabal]: What about 2.2 specifically? And why no >=?

File: src/hevm/src/EVM.hs [hunk 3]
```diff
@@ -1027,53 +1030,174 @@ exec1 = do
...
+
+        -- IDENTITY
+        0x4 -> do
+            assign (state . stack) (1 : xs)
+            assign (state . returndata) input
+            copyBytesToMemory input inSize 0 outOffset
+            next
+
+        -- EXPMOD
+        0x5 -> notImplemented
```
Comment (rainbreak) [inline src/hevm/src/EVM.hs]: When implemented?

File: src/hevm/src/EVM.hs [hunk 4]
```diff
@@ -1895,19 +2023,21 @@ mkCodeOps bytes = RegularVector.fromList . toList $ go 0 bytes
 -- Gas cost function for CALL, transliterated from the Yellow Paper.
 costOfCall
   :: FeeSchedule Word
-  -> Maybe a -> Word -> Word -> Word
+  -> Bool -> Word -> Word -> Word
   -> (Word, Word)
-costOfCall (FeeSchedule {..}) recipient xValue availableGas xGas =
+costOfCall (FeeSchedule {..}) recipientExists xValue availableGas xGas =
```
Comment (rainbreak) [inline src/hevm/src/EVM.hs]: What was wrong with Maybe?

**Comment by rainbreak** [comment]
Will wait for +1 from @cmditch. Also would be nice if Travis wasn't broken, but that's unrelated.

File: src/hevm/hevm.cabal (hunk 1 above)
Comment (livnev) [reply src/hevm/hevm.cabal]: not sure, I guess the cabal file was regenerated in commit 27dbc97c0bbba34907c115abf5e9cea4821fa019

File: src/hevm/hevm.cabal (hunk 2 above)
Comment (livnev) [reply src/hevm/hevm.cabal]: see 27dbc97c0bbba34907c115abf5e9cea4821fa019. We could probably revert the cabal file to how it was, leaving just the `libff` related changes.

File: src/hevm/src/EVM.hs (hunk 4 above)
Comment (livnev) [reply src/hevm/src/EVM.hs]: `costOfCall` used to take a `Maybe Contract` but only checked `isNothing`, I think this was sloppy (and sometimes burdensome to actually produce a term of that type) so I changed it to take a `Bool`.

**Comment by livnev** [comment]
> What's the before/after test suite score?

@rainbreak the last change that affected the VM was merged in #145. Note that this PR not only implements the missing precompiles, but also:

(i) @cmditch fixed a bug with call gas costs
(ii) I fixed the handling of selfdestruct during initcode
(iii) There were some `BlockchainTests` which we were failing to convert into actual test cases. For instance, some of them had no transactions to test. I enquired about the meaning of these, and determined that they were safe to skip, so they are now marked as `ok`.

So we have gone from:
5080 passed, 4000 bad balance, 1532 failed
to
6015 passed, 3782 bad-balance, 815 failed

File: src/hevm/src/EVM.hs (hunk 3 above)
Comment (livnev) [reply src/hevm/src/EVM.hs]: next weekend 😹

**Comment by livnev** [comment]
> Will wait for +1 from @cmditch. Also would be nice if Travis wasn't broken, but that's unrelated.

Actually Travis was reporting a relevant failure: one of the commits that changed `haskell.nix` was bad. I have fixed it and force pushed.

**Comment by livnev** [comment]
@rainbreak also worth warning that [this issue](https://github.com/scipr-lab/libff/issues/15) will mean that it won't build on older MacOS. Do you think we should do something about that?

**Comment by livnev** [comment]
```
  ecrecover
    Example 1:                Query <EVM.Query: fetch contract 0000000000000000000000000000000000000001>
FAIL (0.03s)
      test/test.hs:206:
      bytes32 h = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", d));
      x = ecrecover(h, a, b, c);
      
      expected: Just [0,0,0,0,0,0,0,0,0,0,0,0,45,94,86,212,92,99,21,13,147,127,33,130,83,138,15,24,81,12,177,31]
       but got: Nothing
```
curious, it looks like this has started to fail for some reason, I'll investigate. I wasn't running these tests locally before.

File: src/hevm/src/EVM.hs [hunk 5]
```diff
@@ -1027,53 +1030,173 @@ exec1 = do
...
+  => VM
+  -> FeeSchedule Word
+  -> Word
+  -> Addr
+  -> Addr
+  -> Word
+  -> Word -> Word -> Word -> Word
+  -> [Word]
+  -> EVM ()
+precompiledContract vm fees gasCap precompileAddr recipient xValue inOffset inSize outOffset outSize xs =
```
Comment (cmditch) [inline src/hevm/src/EVM.hs]: Good call parameterizing everything, better without that previous case statement for the call type imo.

**Comment by cmditch** [comment]
@livnev Getting some build errors on the C parts in macOS.
Error output: https://gist.github.com/cmditch/bff97daf634e5f1c6b0ac75d4bd6775a

**Comment by livnev** [comment]
@cmditch I think I finally got it to build on macOS (on travis). The secret was to use `cxx-options` I think. Let me know if it now builds on your machine as well.

**Comment by cmditch** [comment]
Compilation works!

I noticed a regression in the tests that send eth to the precompile address, e.g. this test used to pass:
image

As you said above, you changed this here https://github.com/dapphub/dapptools/commit/b9a6025f68b6a8a2a95cf9554fca504211fbc42a in order to get the Haskell tests working.

**Comment by livnev** [comment]
@cmditch Good to hear that it builds.

I think it's right that it's marked as `bad-balance`: `bad-balance` is for when the states are equal but only if you ignore the balances. Thanks for spotting this regression, of course `touchAccount` is actually the right thing to do, otherwise the ether sent just goes missing since the recipient account never gets initialised. I am considering putting it back in and instead just removing that Haskell test, since I haven't been able to figure out how to properly initialise the state in those tests yet.

**Comment by livnev** [comment]
@rainbreak imo this can be merged now.

**Comment by nanexcool** [comment]
This PR is amazing :heart:

**Comment by rainbreak** [comment]
@cmditch does `stack build` still work for you with this?

**Comment by livnev** [comment]
This now includes the last missing precompile, `MODEXP`.

**Comment by rainbreak** [comment]
Merging. Thanks everyone :) 

@livnev can you make a new issue for the remaining modexp failures?

**Pull request merged**